hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/rollback/ListingBasedRollbackStrategy.java

```diff
@@ -221,49 +216,61 @@ private static String formatDeletePath(String path) {
     if (isCommitMetadataCompleted && tableType == HoodieTableType.COPY_ON_WRITE) {
       return fetchFilesFromCommitMetadata(instantToRollback, partitionPath, basePath, commitMetadataOptional.get(),
-          baseFileExtension, fs);
+          baseFileExtension, storage);
     } else {
-      return fetchFilesFromListFiles(instantToRollback, partitionPath, basePath, baseFileExtension, fs);
+      return fetchFilesFromListFiles(instantToRollback, partitionPath, basePath, baseFileExtension, storage);
     }
   }
 
-  private FileStatus[] fetchFilesFromCommitMetadata(HoodieInstant instantToRollback, String partitionPath,
-                                                    String basePath, HoodieCommitMetadata commitMetadata,
-                                                    String baseFileExtension, FileSystem fs)
-      throws IOException {
-    SerializablePathFilter pathFilter = getSerializablePathFilter(baseFileExtension, instantToRollback.getTimestamp());
-    Path[] filePaths = getFilesFromCommitMetadata(basePath, commitMetadata, partitionPath);
-
-    return fs.listStatus(Arrays.stream(filePaths).filter(entry -> {
-      try {
-        return fs.exists(entry);
-      } catch (Exception e) {
-        LOG.error("Exists check failed for " + entry.toString(), e);
-      }
-      // if any Exception is thrown, do not ignore. let's try to add the file of interest to be deleted. we can't miss any files to be rolled back.
-      return true;
-    }).toArray(Path[]::new), pathFilter);
+  private List<StoragePathInfo> fetchFilesFromCommitMetadata(HoodieInstant instantToRollback,
+                                                             String partitionPath,
+                                                             String basePath,
+                                                             HoodieCommitMetadata commitMetadata,
+                                                             String baseFileExtension,
+                                                             HoodieStorage storage) {
+    StoragePathFilter pathFilter = getPathFilter(baseFileExtension,
+        instantToRollback.getTimestamp());
+    List<StoragePath> filePaths = getFilesFromCommitMetadata(basePath, commitMetadata, partitionPath)
+        .filter(entry -> {
+          try {
+            return storage.exists(entry);
+          } catch (Exception e) {
+            LOG.error("Exists check failed for " + entry.toString(), e);
+          }
+          // if any Exception is thrown, do not ignore. let's try to add the file of interest to be deleted. we can't miss any files to be rolled back.
+          return true;
+        }).collect(Collectors.toList());
+    try {
+      return storage.listDirectEntries(filePaths, pathFilter);
+    } catch (IOException ioe) {
+      LOG.error("Failed to get StoragePathInfo", ioe);
+    }
+
+    return new ArrayList<>();
```

Comment:
Should this still throw `IOException` as before?

Reply:
Yes, good point